### ISSUE TITLE
Celery workers stop consuming tasks after a certain amount of time

### ISSUE THREAD

**Issue by RishabhVerma** [celery/celery#2296]
We have recently deployed a celery cluster with 3 workers on Google Compute Engine.

The celery workers stop consuming tasks after a certain amount of time (around 30 minutes to 1 hour) of running. Stopping the workers and starting them again makes them queue tasks again.

What would be the possible issue for the same?

**Comment by ErikEvenson** [comment]
I believe I am having this problem on heroku with Redis as well: http://stackoverflow.com/q/30038774/382374

**Comment by AdrianLC** [comment]
We might be experiencing the same issue. Has anyone found a solution?

Running on EC2 with RabbitMQ broker and redis result backend.
Strace for more info:
```
epoll_ctl(3, EPOLL_CTL_DEL, 5, {EPOLLWRNORM|EPOLLWRBAND|EPOLLMSG|EPOLLERR|EPOLLHUP|EPOLLONESHOT|0xc698820, {u32=32667, u64=22396489217114011}}) = -1 ENOENT (No such file or directory)
gettimeofday({1432131047, 936437}, NULL) = 0
epoll_wait(3, {{EPOLLIN, {u32=9, u64=39432335262744585}}}, 1023, 1590) = 1
fcntl(9, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
recvfrom(9, "\1\0\3\0\0\0\36", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0<\0013\0\0\0\0\0\0\0y\0\rcelery.pidbox\0", 30, 0, NULL, NULL) = 30
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\2\0\3\0\0\0G", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0\0\0\0\0\0\0\0\0A\370\0\20application/json\5utf-8\0\0\0\34\7expiresdA\325W%zrR\271\5clockI\0-\2017\2\0", 71, 0, NULL, NULL) = 71
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\3\0\3\0\0\0A", 7, 0, NULL, NULL) = 7
recvfrom(9, "{\"destination\": null, \"method\": \"enable_events\", \"arguments\": {}}", 65, 0, NULL, NULL) = 65
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
fcntl(9, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
fcntl(9, F_SETFL, O_RDWR)               = 0
epoll_ctl(3, EPOLL_CTL_DEL, 5, {EPOLLWRNORM|EPOLLWRBAND|EPOLLMSG|EPOLLERR|EPOLLHUP|EPOLLONESHOT|0xc698820, {u32=32667, u64=22396489217114011}}) = -1 ENOENT (No such file or directory)
fcntl(9, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
recvfrom(9, 0x7f9b3be3acb4, 7, 0, 0, 0) = -1 EAGAIN (Resource temporarily unavailable)
fcntl(9, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
fcntl(9, F_SETFL, O_RDWR)               = 0
gettimeofday({1432131048, 788558}, NULL) = 0
epoll_wait(3, {{EPOLLIN, {u32=9, u64=39432335262744585}}}, 1023, 737) = 1
fcntl(9, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
recvfrom(9, "\1\0\2\0\0\0)", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0<\0012\0\0\0\0\0\0\4\374\0\10celeryev\20worker.heartbeat", 41, 0, NULL, NULL) = 41
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\2\0\2\0\0\0]", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0\0\0\0\0\0\0\0\1,\370\0\20application/json\5utf-8\0\0\0002\10hostnameS\0\0\0$celery@albert_replicas.ip-10-0-2-187\2\0", 93, 0, NULL, NULL) = 93
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\3\0\2\0\0\1,", 7, 0, NULL, NULL) = 7
recvfrom(9, "{\"sw_sys\": \"Linux\", \"clock\": 13720503, \"timestamp\": 1432131049.347757, \"hostname\": \"celery@albert_replicas.ip-10-0-2-187\", \"pid\": 19389, \"sw_ver\": \"3.1.18\", \"utcoffset\": 0, \"loadavg\": [0.2, 0.17, 0.15], \"processed\": 3084, \"active\": 0, \"freq\": 2.0, \"type\": \"worker-heartbeat\", \"sw_ident\": \"py-celery\"}", 300, 0, NULL, NULL) = 300
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
fcntl(9, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
fcntl(9, F_SETFL, O_RDWR)               = 0
gettimeofday({1432131049, 350190}, NULL) = 0
epoll_ctl(3, EPOLL_CTL_DEL, 5, {EPOLLWRNORM|EPOLLWRBAND|EPOLLMSG|EPOLLERR|EPOLLHUP|EPOLLONESHOT|0xc698820, {u32=32667, u64=22396489217114011}}) = -1 ENOENT (No such file or directory)
fcntl(9, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
recvfrom(9, "\1\0\2\0\0\0)", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0<\0012\0\0\0\0\0\0\4\375\0\10celeryev\20worker.heartbeat", 41, 0, NULL, NULL) = 41
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\2\0\2\0\0\0\\", 7, 0, NULL, NULL) = 7
recvfrom(9, "\0<\0\0\0\0\0\0\0\0\1(\370\0\20application/json\5utf-8\0\0\0001\10hostnameS\0\0\0#celery@albert_metrics.ip-10-0-2-187\2\0", 92, 0, NULL, NULL) = 92
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
recvfrom(9, "\3\0\2\0\0\1(", 7, 0, NULL, NULL) = 7
recvfrom(9, "{\"sw_sys\": \"Linux\", \"clock\": 13720503, \"timestamp\": 1432131049.348449, \"hostname\": \"celery@albert_metrics.ip-10-0-2-187\", \"pid\": 19388, \"sw_ver\": \"3.1.18\", \"utcoffset\": 0, \"loadavg\": [0.2, 0.17, 0.15], \"processed\": 4, \"active\": 0, \"freq\": 2.0, \"type\": \"worker-heartbeat\", \"sw_ident\": \"py-celery\"}", 296, 0, NULL, NULL) = 296
recvfrom(9, "\316", 1, 0, NULL, NULL)   = 1
fcntl(9, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
fcntl(9, F_SETFL, O_RDWR)               = 0
gettimeofday({1432131049, 350899}, NULL) = 0
fcntl(9, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(9, F_SETFL, O_RDWR|O_NONBLOCK)    = 0
recvfrom(9, 0x7f9b3be3a054, 7, 0, 0, 0) = -1 EAGAIN (Resource temporarily unavailable)
fcntl(9, F_GETFL)                       = 0x802 (flags O_RDWR|O_NONBLOCK)
fcntl(9, F_SETFL, O_RDWR)               = 0
gettimeofday({1432131049, 351125}, NULL) = 0
```

**Comment by mwaaas** [comment]
same issue in rabbitmq    but happens after several days

**Comment by thedrow** [comment]
We really need to reproduce this.
I have been experiencing the same with RabbitMQ.
Do any of you have a failing test case?

**Comment by RishabhVerma** [comment]
No I don't. In fact I was facing this issue when working with redis but it
vanished after moving to RabbitMQ.

**Comment by bryanhelmig** [comment]
Same thing for us, the worker hangs (perhaps after a VM pause or network blip?), these logs are all that exist:

```
[2015-09-24 15:36:17,812: INFO/MainProcess] missed heartbeat from celery@ip-10-0-123-123
[2015-09-24 15:37:20,087: INFO/MainProcess] sync with celery@ip-10-0-234-234
```

We're still digging in and trying to discover what is going on - we're RabbitMQ and Celery 3.1.18 - results disabled. Gossip/heartbeat/etc. all still enabled (we hit some possibly related bugs like https://github.com/celery/celery/issues/1847 trying to turn them off).

**Comment by rafaelosoto** [comment]
I'm seeing the same behavior with RabbitMQ 3.2.4 and Celery 3.1.18. The workers were running fine for a few months the first time, then after a restart it was a few weeks till it happened again. Logs also only show missed heartbeats from the worker that stopped.

**Comment by bryanhelmig** [comment]
@rafaelosoto our "hangs" were a mix of a couple things over time:

* Disabling gossip/heartbeat/mingle/etc. can see some nasty startup bugs (prefetching to the limit but never consuming the messages) - happens when queue is deep and workers startup.
* Too much load on the haproxy boxes w/ single core between celery/rabbit (loooong syncs for a big cluster). 
* A kernel bug "skb rides the rocket error" incorrectly hanging AMQP connections under load (specifically on the haproxy boxes).

Just in case folks are hitting similar problems - ^^ was a lot of suffering to resolve them all.

**Comment by flsafe** [comment]
We think we might be seeing a similar issue with celery 3.1.17 and rabbit 3.5.0.

We've noticed missed heart beats, and time drift related messages in our celery logs. When we attached strace to the celery process we noticed a similar pattern of spinning on `epoll` and `recvfrom` calls.

**Comment by rafaelosoto** [comment]
@bryanhelmig thanks a lot for the response! We haven't disabled gossip/heartbeats/mingle so far but I'll take a look at the other issues you mentioned.

**Comment by mu4farooqi** [comment]
I'm also facing this issue with Celery 3.1.18 and Redis. Any solution ?

**Comment by ccdesales** [comment]
My workers were having the same issue. I noticed that the longer they were running, the more memory usage was, until the whole memory was overloaded; that's when they started being non responsive. 

Using a low value for CELERY_MAX_TASKS_PER_CHILD made things much better.

**Comment by ask** [comment]
This seems to be a meta-issue, lots of different people experiencing a very generic symptom, but probably they have little in common.

Redis, Rabbitmq, epoll, out of memory,

Redis issues could be resolved with the recent deadlock fixes (in 3.1.22)
epoll issue could be fixed with the recent 100% CPU with epoll fix (in 3.1.22)

I think we need to close this issue, or it will just stay open indefinitely collecting anecdotes from people
with unrelated issues.

If anyone keep seeing this symptom after upgrading to 3.1.22, then please open up a new separate
issue, and be sure to include the configuration that you use (broker, result backend, worker command-line arguments, worker configuration)

E.g. if you're using redis as the broker, then please add that to the title of the issue.

We want to help you fix the problems, it's just that we would also like to keep track of what issues are resolved, and that we are asking the right questions based on your configuration.

**Comment by thedrow** [comment]
:+1:

**Comment by coderonline** [comment]
I am getting this error when I accidentally run the celery daemon multiple times

**Comment by thedrow** [comment]
@coderonline Please open a new issue about it.

**Comment by yeago** [comment]
See also #4185

**Comment by alugowski** [comment]
I'm also experiencing this problem. I'll add some observations here instead of starting a new issue because I don't expect anyone to pick this up and Googling brings up this page.

The trend I detect here is large parallelism and/or worker count. It's certainly true in my case. My workflow is about 2,000 hour-long tasks, each using 8 cores, celery concurrency=1. I'm running in Kubernetes Engine. I've been seeing these dead workers about every three hours. Some of the tricks above appear to have slightly helped, though I can't tell you which ones since the settings appear to be ignored. What probably helped the most is to prevent too many workers from starting at the same time and capping total concurrent workers.

I've tried several Celery versions, using both RabbitMQ (several versions) and Redis. Redis lasts longer for me. I had assumed the problem is RabbitMQ because despite the docs saying it's the recommended backend the startup message says its deprecated. I'm now on Celery 4.2 with Redis 4.

Dead workers show missed heartbeats and failed connections to Redis. Once one worker dies they all follow as they run out of tasks. I also have Flower in the same cluster and it continues to function normally. I can watch workers run out of tasks and become idle in real time.

To me this smells like a race condition. I've not delved into the service discovery and heartbeat code, but this sounds like corruption of some datastructure there. Maybe concurrent modification when the GIL is released due to a library call? I think a kludge fix would be to restart networking when the broker reconnect attempt counter drops to zero.

I suspect the way to reproduce this issue it is to build a high CPU, high parallelism workflow in containers and keep running it. It's an intermittent issue and those are always a challenge.

**Comment by babu-upcomer** [comment]
@alugowski You spoke my mind. Just ended up with the same scenario today. Workers hung up due to missing heartbeats from other celery workers. Flower showed every workers online though!

Celery 4
Gevent workers
Redis 4?(Elasticcache)

**Comment by alugowski** [comment]
@babu-upcomer I created an issue with more details: #5037

If that also matches what you're seeing then feel free to chime in there so it doesn't look like just a problem one abnormal user is seeing.

**Comment by bakcsa83** [comment]
Hi All,
I just started to use Celery and I immediately ran into this issue. I can reproduce it very quickly with a simple script (tasks.py see attached.)
It always hangs at line 33 `results.join()` between the 50th and 400th iteration of the for loop in the `start()` function.
Worker started with the following command:
`celery -A tasks worker -l info`
Broker: RabbitMQ 3.6.6
Backend: Redis 4.0.11
OS: Debian 9 (Running in Virtualbox with 4 cores and 8G ram)
You can find all logs attached but only rabbitmq.log contains errors.
[rabbitmq.log](https://github.com/celery/celery/files/2358902/rabbitmq.log)
[redis_log.txt](https://github.com/celery/celery/files/2358903/redis_log.txt)
[requirements.txt](https://github.com/celery/celery/files/2358904/requirements.txt)
[script output.txt](https://github.com/celery/celery/files/2358905/script.output.txt)
[worker_output.txt](https://github.com/celery/celery/files/2358906/worker_output.txt)

**Comment by alugowski** [comment]
@bakcsa83 open an issue. If you have a reproducible case then that should be solvable. Note that I think you forgot to attach tasks.py

**Comment by bakcsa83** [comment]
@alugowski I've got the feeling that my problem is not the same as this one. I've opened another issue where I posted a shorter version of the script. I hope somebody can either confirm that there is indeed a bug in Celery or tell me what's wrong with my script. #5040

**Comment by junbujianwpl** [comment]
Any solution yet?  I met the same problem. 
Python==3.6
celery==4.1.1
Django==2.0.6
RabbitMQ=3.1.5
When it comes, `celery inspect active` returns nothing. Celery beat works good and I can check it from the log.  But celery worker log stopped several days ago. RabbitMq running good. `Celery inspect registered` is good. Celery processes are good and I can check  them with `ps` command.

**Comment by yeago** [comment]
I patched this according to  https://github.com/celery/celery/issues/4185 and
never encountered it again

**Comment by justquick** [comment]
Had the same issue and removing gevent as the pool implementation solved it.

Python 3.8.7
Celery 4.4.7
RabbitMQ 3.8.14

**Comment by radiantone** [comment]
> Had the same issue and removing gevent as the pool implementation solved it.
> 
> Python 3.8.7
> Celery 4.4.7
> RabbitMQ 3.8.14

And how do you do that exactly?

**Comment by DevanshShah2002** [comment]
Hey I am facing same kind of problem. when I run my code it is working properly but after running for 70/80 mins without error or warning suddenly my threads are showing succeed and my celery stop execution. then when I restart or check queue there is still pending tasks in that queue... Any solution there? may be its look like timeout for worker is it can change?

**Comment by Anko7** [comment]
In our case changing the `prefetch-multiplier` setting from 4 to 1 helped.

**Comment by darayn** [comment]
I was facing same issue with my celery worker. when restarted along with redis, celery work used to stop consuming tasks in 20mins approx and if i just restart the celery worker it would resume pending task.

Did tried all the solutions above. No Luck. 

Then decided to upgrade my celery and redis version, still saw same issue coming again after a month. 

When I checked my upgrade scripts I was cleaning up the redis dump, this caught my eye on it. 

after reproducing this issue multple times tried cleaning up the redis dump and appendonly memory on location `/var/lib/redis` and restarting both. And the issue did disappear like it was never there. 

I was seeing this issue for the first time only after few weeks after I added persistence to redis config.
